R/parse-body.R

```diff
@@ -159,24 +180,34 @@ parser_json <- function(...) {
 #' @export
 parser_text <- function(parse_fn = identity) {
   stopifnot(is.function(parse_fn))
-  function(value, content_type = NULL, ...) {
+  parse_func <- function(value, content_type = NULL, ...) {
     charset <- getCharacterSet(content_type)
     value <- rawToChar(value)
     Encoding(value) <- charset
     parse_fn(value)
   }
+  return(invisible(
+    list("text/plain" = parse_func,
+         "text" = parse_func)
+  ))
 }
 
 
 #' @describeIn parsers YAML parser
 #' @export
 parser_yaml <- function(...) {
-  parser_text(function(val) {
+  parse_func <- parser_text(function(val) {
     if (!requireNamespace("yaml", quietly = TRUE)) {
       stop("yaml must be installed for the yaml parser to work")
     }
     yaml::yaml.load(val, ...)
-  })
+  })[[1]]
+  return(invisible(
+    list("application/yaml" = parse_func,
+         "application/x-yaml" = parse_func,
+         "text/yaml" = parse_func,
+         "text/x-yaml" = parse_func)
+  ))
```

Comment:
I'm on the fence as to which approach I'd for `plumber` to use

One side says we should have a fixed set of values. The other side says we should have a regex.

1. Fixed
```r
list(
  content_types = c("application/yaml", "application/x-yaml", "text/yaml", "text/x-yaml"),
  parser = parse_func
)
```
* Pro: O(1) lookup
* Con: Does **not** allow for dynamic values

2. Regex
```r
list(
  content_type_regex = "^(application|text)/(x-)?yaml$",
  parser = parse_func
)
```
* Con: O(n) look up. Slower. Must test each parser `content_type_regex`. (`n` isn't _that_ big though. Maybe up to 5 values regularly?)
* Pro: Allows for dynamic/partial matching content types

----------------------

Typing this out, I think we should change to using a regex and function combination. Since the content type is passed into the parsing function, the function can handle any particular differences at that point.

Should we use a helper function to validate what is being returned? Such as `plumber_parser(content_type_regex, parser)` or `init_parser(content_type_regex, parser)`? (I think `parser()` would be too generic.

Does this sound ok?

Reply:
I'll test if the way it is set up at the moment cause any performance decay. I agree it is not clean.

Reply:
So using a single regex :
```R
microbenchmark::microbenchmark(stri_detect_regex("text/html", "(application|text)/(plain|html)"))
Unit: microseconds
                                                              expr   min
 stri_detect_regex("text/html", "(application|text)/(plain|html)") 7.501
    lq    mean median    uq    max neval
 7.701 9.07805  7.802 8.901 42.101   100
```

O(1) lookup using the current list
```R
> parsers <- parser_all()
> microbenchmark::microbenchmark(parsers$text)
Unit: nanoseconds
         expr min  lq   mean median  uq   max neval
 parsers$text 300 301 514.98    301 400 18601   100
> microbenchmark::microbenchmark(parsers$`application/json`)
Unit: nanoseconds
                       expr min  lq   mean median  uq  max neval
 parsers$`application/json` 100 200 258.98    201 201 5200   100
```

From a performance perspective, I would stick with a non regex solution.

Reply:
@jcheng5 There is a minimum 10x performance penalty for using a regex when validating a parser works for a given content type. As more parsers are added, more parsers would need to be tested, slowing down even further. 

Should we use a slower regex approach to gain regex matching capabilities or should we stay with the faster, white-list approach?

---------

Notes: Units for regex testing are `microseconds`. Units for key/value lookup (`parser_all()`) are `nanoseconds`.

Reply:
That's still pretty fast. Can you really get away with exact matching? e.g. `application/json; charset=utf-8` is pretty common; the space after the semicolon is optional and the parameter name is case insensitive. I'm not saying you have to support all of that kind of stuff out of the gate, but I have a feeling either you or plumber users will need to deal with it eventually.

Reply:
we only match on mime-type, charset is handled inside individual parser.

https://github.com/rstudio/plumber/blob/0d3b795949b892b2f5efadc6abf5d8f59a96f105/R/parse-body.R#L48

Reply:
Oh OK, that takes care of a lot then. The only other thing I can think of is how XML has subtypes, like `application/atom+xml`, `application/xhtml+xml`. I don't have an opinion which you should choose but, if you choose fixed strings, I'd suggest you have an idea how you would introduce something more dynamic in the future if necessary (without breaking compatibility). Which maybe argues for the helper function approach.

Reply:
I will check but I think all headers are lower cased when they reach us.
I also believe header value to be case sensitive per. Only the name is case insensitive. 
One exception is charset https://tools.ietf.org/html/rfc2616#section-3.4 but we do not use it. 

https://stackoverflow.com/questions/5258977/are-http-headers-case-sensitive
https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2

Reply:
> Oh OK, that takes care of a lot then. The only other thing I can think of is how XML has subtypes, like `application/atom+xml`, `application/xhtml+xml`. I don't have an opinion which you should choose but, if you choose fixed strings, I'd suggest you have an idea how you would introduce something more dynamic in the future if necessary (without breaking compatibility). Which maybe argues for the helper function approach.

I'll see if we can up with something cuter.

For now it would be something like
```R
parser_xml <- function() {
  f <- parser_text(mockxmlparserfunction)
  return(list(invisible(
    "application/atom+xml" = f,
    "application/xhtml+xml" = f,
    "application/xml" = f,
    "application/xhtml" = f
  )))
add_parser("xml", parser_xml)
```

Then you could it 
```R
#* @parser xml
```

Reply:
If we had `parsers` be the list of endpoint parsers that contain fixed content types and/or a regex, we could iterate through them.



```
for (parser_info in parsers) {
  parser_content_types <- parser_info$content_types
  if (!is.null(parser_content_types)) {
    if (content_type %in% parser_content_types) {
      return(parser_info$parser)
    } 
  }
  parser_content_type_regex <- parser_info$content_type_regex
  if (!is.null(parser_content_type_regex)) {
    if (stri_detect(content_type, regex = parser_content_type_regex)) {
      return(parser_info$parser)
    }
  }
}
```

Right now, everything could be in the top `parser_content_types` set.  But it would also allow for the regex within xml. 

This would also allow for multiple parsers to have the same content type, so order of the parsers would matter.  

(`parser_text()` could also add a regex for `text/` in addition to the known type of `text/plain`)

Reply:
When would you execute this loop?

Reply:
Np! I should be more explicit. Sorry about that

---------------------

I would replace this line with the loop above: 

https://github.com/rstudio/plumber/pull/587/files#diff-82431f6d2dc55b4007eff34ac0a9d103R60

Reply:
@schloerke 

I ended up doing something similar but different, but similar. I kept the fixed match, if nothing is found, it does a first regex match. Hit me up on zoom if you want to discuss anything or here.

Ready for review, tests and parsers doc done.

Reply:
@meztez Sent you an email with a link

Reply:
Example above now become:
```R
parser_xml <- function() {
  f <- parser_text(mockxmlparserfunction)
  ct <- c("application/xhtml+xml",
          "(xml|xhtml)$")
  return(make_parsers(f, ct))
}
add_parser("xml", parser_xml)
```

Reply:
And before you suggest it. Do you think we should put `ct` as a parameter to the parser builder like so:

```R
parser_xml <- function(match_type =  c("application/xhtml+xml", "(xml|xhtml)$")) {
  f <- parser_text(mockxmlparserfunction)
  return(make_parsers(f, match_type))
}
add_parser("xml", parser_xml)
```

This way it could be exposed, better documented and usable directly from the `@parser` tag.